src/exp_and_gram.jl

```diff
@@ -17,70 +17,129 @@ struct ExpAndGram{T,N,A,B} <: AbstractExpAndGramAlgorithm where {T,N,A,B}
-    return Φ, G
-end
+exp_and_gram(A::AbstractMatrix, B::AbstractMatrix, method::AbstractExpAndGramAlgorithm) =
+    exp_and_gram(A, B, true, method)
+exp_and_gram(A::AbstractMatrix, B::AbstractMatrix, t::Number, method::AbstractExpAndGramAlgorithm) =
+    exp_and_gram!(similar(A), similar(A), A, B, t, method)
+
+exp_and_gram!(
+    eA::AbstractMatrix,
+    U::AbstractMatrix,
+    A::AbstractMatrix,
+    B::AbstractMatrix,
+    method::AbstractExpAndGramAlgorithm,
+    cache = alloc_mem(A, B, method),
+) = exp_and_gram!(eA, U, A, B, true, method, cache)
 
 function exp_and_gram!(
-    A::AbstractMatrix{T},
-    B::AbstractMatrix{T},
-    method::ExpAndGram,
-) where {T<:Number}
-    Φ, U = exp_and_gram_chol!(A, B, method)
+    eA::AbstractMatrix,
+    U::AbstractMatrix,
+    A::AbstractMatrix,
+    B::AbstractMatrix,
+    t::Number,
+    method::AbstractExpAndGramAlgorithm,
+    cache = alloc_mem(A, B, method),
+)
+    eA, U = exp_and_gram_chol!(eA, U, A, B, t, method, cache)
     G = U' * U
     _symmetrize!(G)
-    return Φ, G
+    return eA, G
 end
 
-exp_and_gram_chol(
-    A::AbstractMatrix{T},
-    B::AbstractMatrix{T},
-    method::ExpAndGram{T},
-) where {T<:Number} = exp_and_gram_chol!(copy(A), copy(B), method)
+function alloc_mem(A, B, method::ExpAndGram{T,q}) where {T,q}
+    n, m = size(B)
+    # return (
+    #     U = similar(A),
+    #     pre_array = similar(A, 2n, n),
+    #     tmp = similar(A),
+    # )
+    return nothing
+end
+function alloc_mem(A, B, method::ExpAndGram{T,13}) where {T}
+    q = 13
+    n, m = size(B)
+    return (
+        _A = similar(A),
+        _B = similar(B),
+        A2 = similar(A),
+        A4 = similar(A),
+        A6 = similar(A),
+        tmpA1 = similar(A),
+        tmpA2 = similar(A),
+        tmpA3 = similar(A),
+        L = similar(A, n, m * (q + 1)),
+        tmpB2 = similar(B),
+        A2B = similar(B),
+        A4B = similar(B),
+        A6B = similar(B),
+        U = similar(A),
+        pre_array = similar(A, 2n, n),
+        tmp = similar(A),
+    )
+end
+
+exp_and_gram_chol(A::AbstractMatrix, B::AbstractMatrix, method::AbstractExpAndGramAlgorithm, cache = alloc_mem(A, B, method)) =
+    exp_and_gram_chol!(similar(A), similar(A), A, B, method, cache)
+exp_and_gram_chol(A::AbstractMatrix, B::AbstractMatrix, t::Real, method::AbstractExpAndGramAlgorithm, cache = alloc_mem(A, B, method)) =
+    exp_and_gram_chol!(similar(A), similar(A), A, B, t, method, cache)
+
+exp_and_gram_chol!(eA::AbstractMatrix, U::AbstractMatrix, A::AbstractMatrix, B::AbstractMatrix, method::AbstractExpAndGramAlgorithm, cache = alloc_mem(A, B, method)) =
+    exp_and_gram_chol!(eA, U, A, B, true, method, cache)
 
 function exp_and_gram_chol!(
+    eA::AbstractMatrix{T},
+    U::AbstractMatrix{T},
     A::AbstractMatrix{T},
     B::AbstractMatrix{T},
+    t::Real,
     method::ExpAndGram{T,q},
-) where {T<:Number,q}
+    cache = alloc_mem(A, B, method),
+) where {T,q}
+    At, Bt = if cache == nothing
```

Comment:
This part is still temporary, I'll change this once the other method also uses a cache